File: leetcode/51-n-queens.cpp
/*
The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens attack each other.



Given an integer n, return all distinct solutions to the n-queens puzzle.

Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both indicate a queen and an empty space respectively.

For example,
There exist two distinct solutions to the 4-queens puzzle:

[
 [".Q..",  // Solution 1
  "...Q",
  "Q...",
  "..Q."],

 ["..Q.",  // Solution 2
  "Q...",
  "...Q",
  ".Q.."]
]
*/

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<int> row(n, 0);
        set<int> available;
        for (int i=0; i<n; ++i) {
            available.insert(i);
        }
        vector<vector<int>> solutions;
        step(row, available, 0, solutions);
        vector<vector<string>> results;
        string line(n, '.');
        for (auto& solution : solutions) {
            vector<string> result;
            for (auto i : solution) {
                auto tmp = line;
                tmp[i] = 'Q';
                result.push_back(move(tmp));
            }
            results.push_back(move(result));
        }
        return results;
    }

    void step(vector<int> &row, const set<int> &available,
              int next, vector<vector<int>> &solutions) {
        if (next >= row.size()) {
            return;
        }
        for (auto candidate : available) {
            auto variable = true;
            for (auto j = 0; j < next; ++j) {
                if (next - j == abs(candidate - row[j])) {
                    variable = false;
                    break;
                }
            }
            if (variable) {
                row[next] = candidate;
                if (next + 1 == row.size()) {
                    solutions.push_back(row);
                } else {
                    set<int> tmp_available = available;
                    tmp_available.erase(candidate);
                    step(row, tmp_available, next+1, solutions);
                }
            }
        }
    }
};

File: leetcode/51-n-queens.rb
# @param {Integer} n
# @return {String[][]}
def solve_n_queens(n)
  $sols = []  
  cols = Array.new n.size, false
  x1 = Array.new n.size*2, false
  x2 = Array.new n.size*2, false
  board = n.times.map { '.' * n }
  n_queens n, 0, cols, x1, x2, board
  $sols
end


$sols = nil


def n_queens(n, i, cols, x1, x2, board)
  if n == i
    $sols.push copy board
    return
  end
  (0...n).each do |j|
    next if cols[j]
    x1_idx = i - j + n - 1
    next if x1[x1_idx]
    x2_idx = i + j
    next if x2[x2_idx]
    cols[j] = true
    x1[x1_idx] = true
    x2[x2_idx] = true
    board[i][j] = 'Q'
    n_queens n, i+1, cols, x1, x2, board
    board[i][j] = '.'
    cols[j] = false
    x1[x1_idx] = false
    x2[x2_idx] = false
  end
end


def copy(board)
  board.map { |row| row.dup }
end


#puts solve_n_queens(4).to_s
